File: Week_2/validParents.js

// var isValid = function(s){
    // var pars 
// }


// s = '()'
// var c = s.split('');
// console.log(c);

var isValid = function(s) {
    var pars = ['(',')','[',']','{','}']
    stack = []
    var c = s.split('')

    // c.forEach((item) =>{
    //     var idx = pars.indexOf(item) 
    //     if(idx % 2 == 0){
    //         stack.push(item)
    //     }else{
    //         if(stack.length == 0){
    //             return false
    //         }else{
    //             val = stack.pop()
    //             if(val != pars[idx -1]){
    //                 return false
    //             }else{
    //                 return true
    //             }
    //         }
    //     }
    // } )

    for(let i =0; i<s.length; i++){
        var idx = pars.indexOf(s[i]);
        if(idx % 2 == 0){
            stack.push(s[i])
        }else{
            if(stack.length == 0){
                return false;
            }else{
                val = stack.pop()
                if(val != pars[idx -1]){
                    return false;
                }else{
                    return true;
                }
            }
        }
    }

    if(stack.lenght === 0){
        return true
    }else{
        return false
    }
    
};

s = '()'
s1 = '()[]{}'
s2 = '(]'

console.log(isValid(s));
console.log(isValid(s1));
console.log(isValid(s2));

File: Week_2/validParents.py
def validParens(s):
    #Runtime Complexity O(N)
    #Space Complexity  O(N) We're adding to our stack depending by N size, so O(N)
    parens = ['(',')','[',']','{', '}']
    stack = []
    #O(N) Iterating N times 
    for c in s:
        #Looking for index of x is O(1)
        idx = parens.index(c)
        if idx % 2 == 0:
            stack += c
        else:
            if len(stack) == 0:
                return False
            else:
                val = stack.pop()
                if val != parens[idx - 1]:
                    return False
                return True
            
    if(len(stack) == 0):
        return True
    return False




if __name__ == '__main__':
    s = "()"
    print(validParens(s))

    s1 = "()[]{}"
    print(validParens(s1))

    s2 = "(]"
    print(validParens(s2))

    s3 = "{[)}"
    print(validParens(s3))

    s4= "{[]}"
    print(validParens(s4))

